### ISSUE TITLE
Deploy failed: not authorized with more than one instance open and adminAuth active in the settings

### ISSUE THREAD

**Issue by thechane** [node-red/node-red#2657]
### What are the steps to reproduce?

1. Bring up a nodered instance that makes use of adminAuth in the settings file. 
2. Log in.
3. Bring up a different nodered instance that makes use of adminAuth in the settings file (but is not the same as the previous one).
4. Log in.
5. Go back to original instance, make a change and hit deploy.

### What happens?
Deploy failed: not authorized (popup) 

### What do you expect to happen?
For it to deploy as normal.

Workaround: duplicate the tab (in Firefox) or bring the original instance up again and log in. Switch back to the original instance and now the deploy works just fine.

### Please tell us about your environment:

- [ ] Node-RED version: 1.0.6
- [ ] Node.js version:  12.16.3
- [ ] npm version: 6.14.4
- [ ] Platform/OS: Alpine Linux (node:12-alpine so alpine:3.11)
- [ ] Browser: Firefox 78.0.2

**Comment by knolleary** [comment]
When you bring up the second instance is it using the same `userDir`? If so, that is not a valid way to run a second instance. Each instance *must* have its own `userDir`.

**Comment by thechane** [comment]
Hello - each instance is running in it's own docker container with projects enabled.

**Comment by knolleary** [comment]
So I assume the second instance you bring up is listening on a different port, with a completely separate `/data` dir (assuming you mount it as a volume to the container)?

**Comment by thechane** [comment]
Yep, - they are completely separated from each other. They do share the same beginnings of their URLs. So for example https://somewhere.com/projectA/admin and https://somewhere.com/projectB/admin

ProjectA and B are completely different in terms of what's in the flow.json.

**Comment by knolleary** [comment]
Ah right - same domain, so the same localStorage scope. Yup, that isn't going to work.

So we need to include the full path somehow in the localStorage scope ... but that poses other security concerns as it would mean the user of projectA would be able to read the localStorage of projectB.

**Comment by knolleary** [comment]
I think the only way to support multiple instances on the same domain is to move away from localStorage and use cookies instead, which can be scoped to the path. Need to work through what that means for the whole way the editor auths with the runtime.

**Comment by thechane** [comment]
Apologies for the can of worms. There is an easy enough workaround, but figured it probably should be looked at.

**Comment by knolleary** [comment]
A good can of worms :)

Most of the multi-instance node-red setups I've seen use subdomains, so this has never come up before. But no reason for us not to support paths on the same domain.

**Comment by knolleary** [comment]
I'm pretty sure we can add in passport-cookie here and switch over to using cookie auth for the editor with no impact to existing users. We'd still support the bearer auth mechanism in place today, but the editor wouldn't need to use it - that would be for automated access to the admin api.

But its late here and this is too big of a change to do at 10pm. Will look at it with fresh eyes tomorrow - but I *think* it'll be a simple enough change.

**Comment by kuema** [comment]
Just a side-note: the same thing happens when you expose multiple instances through one webserver, like an Apache/nginx with reverse proxies to each one on different URI paths.

You could only log into one instance at a time. I already suspected it has something to do with the localstorage, so I worked around it by using private/incognito browser windows.

**Comment by thechane** [comment]
I wonder if a passport-cookie could also be used to resolve the git authentication issue https://github.com/node-red/node-red/issues/2656 ?

**Comment by knolleary** [comment]
@thechane no, its a completely unrelated issue

**Comment by albrecht-j** [comment]
@thechane Hello, do you know when this feature is planned to be finished?
Thanks

**Comment by knolleary** [comment]
Hi @albrecht-j ,

@thechane reported the issue and raised the requirement. It isn't something he is working on.

I had hoped to address it in the 2.0 release, but it slipped out of the plan. I don't have an ETA for the work as it requires a fairly significant change to had auth is handled in the editor. My previous optimism about it being a small change was misplaced - its more complicated than I thought.

**Comment by mw75** [comment]
I just got trapped by this problem. My instances share the same passport auth provider and i have a storageclass attached to them. 

Would it be an option to feed the same sessions.json to the instances? Is there an api-endpoint to request a session reload? 

Can i use https://nodered.org/docs/user-guide/runtime/securing-node-red#custom-authentication-tokens to share authentication? 

Is there a simple option to use the outh2 access token and not creating an internal one?
 
My auth-provider (keycloak) manages sessions on its own and creates new tokens on the auth-loop without user interaction. But it interrupts your workflow when switching between the instances. My instances cover different aspects with different pallets.

**Comment by mw75** [comment]
The approach reads the path from window.location in RED.settings.init(), removes the last character, replaces "/" with "-" and stores it a authTokensSuffix property. Whenever the key auth-tokens is read, set or removed this suffix is added to the localStorage key. This results i no api-changes and no changes when the editor runs in the root path and you may run as much instances as you like on a single host name.

The only, from my perspective minor, point of discussion is the the behavior, that the hole auth-tokens* "Namespace" for setting-keys gets "magic" while currently only the key auth-tokens itself has special meaning. I don't see a security issue because the sessions are initiated from the same operating system user, on the same browser, on the same computer. And as logout is still working as expected, it truly has to be the same person - so nothing is leaked when reviewing the localStorage.  

Bevor filling a PR i have read the contribution guide (again) and i am not quite sure to which branch this PR should point. It applies to dev and master without conflicts!

Feedback is appreciated! Would like to get it merged, because it is needed on our solution and i would like to get rid of that patch while building.

Regards,
Mario
